tests/portability/cmsis_rtos_v2/src/thread_apis.c

```diff
@@ -264,6 +264,15 @@ static void thread4(void *argument)
 	printk(" + Thread A started.\n");
 	status = osThreadJoin(tB);
 	zassert_equal(status, osOK, "osThreadJoin thread B failed!");
+
+	/*
+	 * Both this thread and the main thread (test_thread_join) are
+	 * joining thread5. If this thread is scheduled before the
+	 * main thread this thread will be dead before the main thread
+	 * has chance to join it. Adding a delay (following the test
```

Comment:
If I understand how this test is set up:
- ztest_thread creates Thread B
- ztest_thread creates Thread A
- ztest_thread joins on Thread B, and goes to sleep
- Thread B starts, but it enters an osDelay() of 1000 MS which is a very long time, and goes back to sleep
- Thread A starts, and also joins on Thread B, going to sleep
- Thread B wakes up from its osDelay() call and exits

So is the bug the following scenario:
- Thread A wakes up from its join call on thread B, and exits
- ztest_thread wakes up, tries to join on Thread A, but since it already exited we get an error?

And if ztest_thread wakes up before Thread A, everything works as expected?

In general I haaaaaate fixing issues with test thread scheduling with delays, it's far more robust to use different thread priorities instead. Can we set priorities for Thread A / Thread B so that the expected sequence is guaranteed?

Reply:
That is the scenario, the priorities should be set between thread A and ztest_thread. I agree with you btw, I was explaining it for Aastha, about how bad is syncing with delays, I just followed the way it was being done but let fixes it properly though.

Reply:
Just setting thread priority to the lowest worked, but it seems fragile as well since ztest_thread has some instructions before call thread_join to A and it can be scheduled out in one of those if I understood correctly. IMHO the whole test is confuse and it needs some love. Just sending a new version using priorities right now

Reply:
> has some instructions before call thread_join to A and it can be scheduled out in one of those if I understood correctly

Not sure what you mean by this. What instructions are we talking about which would cause the thread to be scheduled out?

This shouldn't be fragile, the whole execution sequence should be predictable / deterministic

Reply:
ztest_thread calls k_uptime_delta that calls the syscall k_uptime_ticks

Reply:
nevermind, it is not even building with usermode

Reply:
> ztest_thread calls k_uptime_delta that calls the syscall k_uptime_ticks

Unless I'm mistaken neither of those functions invokes the scheduler.

Reply:
does this patch looks right now ?